metrics/metrics.go

```diff
@@ -0,0 +1,163 @@
+	MeasurePeersRefreshed           = "libp2p.io/dht/k-bucket/peers_refreshed"
+	MeasurePeersRemoved             = "libp2p.io/dht/k-bucket/peers_removed"
+)
+
+// Measures are exported so that consumers can create their own views if the `DefaultViews`
+// aren't sufficient. However, they should be updated using the functions below to avoid
+// leaking OpenCensus cruft throughout the rest of the code.
+var (
+	KBucketsFull = stats.Int64(MeasureBucketsFull,
+		"Number of k-buckets that are at capacity (have k entries).", stats.UnitDimensionless)
+	KBucketsNonEmpty = stats.Int64(MeasureBucketsNonEmpty,
+		"Number of k-buckets with at least one entry.", stats.UnitDimensionless)
+	KBucketUtilization = stats.Int64(MeasureBucketUtilization,
+		"Number of entries per k-bucket.", stats.UnitDimensionless)
+	KBucketPeerLatency = stats.Float64(MeasurePeerLatency,
+		"Recorded latency measurements for all peers added to each bucket, per bucket.", stats.UnitMilliseconds)
+	KBucketPeersAdded = stats.Int64(MeasurePeersAdded,
+		"Number of peers added to each k-bucket (cumulative). Note that peers can be counted twice if added and removed.", stats.UnitDimensionless)
+	KBucketPeersRejectedHighLatency = stats.Int64(MeasurePeersRejectedHighLatency,
+		"Number of peers rejected from k-buckets due to high latency.", stats.UnitDimensionless)
+	KBucketPeersRejectedNoCapacity = stats.Int64(MeasurePeersRejectedNoCapacity,
+		"Number of peers rejected from the routing table because their k-bucket was full.", stats.UnitDimensionless)
+	KBucketPeersRefreshed = stats.Int64(MeasurePeersRefreshed,
+		"Number of peers moved to the front of a k-bucket that they were already in.", stats.UnitDimensionless)
+	KBucketPeersRemoved = stats.Int64(MeasurePeersRemoved,
+		"Number of peers removed from each k-bucket (cumulative). "+
+			"Peers may be counted twice if added and removed.", stats.UnitDimensionless)
+)
+
+// LocalContext returns `ctx` tagged with the local dht `id` for metrics reporting.
+func LocalContext(ctx context.Context, id []byte) context.Context {
+	pretty := base58.Encode(id)
+	ctx, _ = tag.New(ctx, tag.Upsert(keyLocalId, pretty))
+	return ctx
+}
+
+// RecordBucketsFull records the current number of buckets `n` that are at capacity.
+func RecordBucketsFull(ctx context.Context, n int) {
+	stats.Record(ctx, KBucketsFull.M(int64(n)))
+}
+
+// RecordBucketsNonEmpty records the current number of buckets `n` that have at least one entry.
+func RecordBucketsNonEmpty(ctx context.Context, n int) {
+	stats.Record(ctx, KBucketsNonEmpty.M(int64(n)))
+}
+
+// recordWithBucketIndex is a helper func that applies a tag to the measurements `ms`
+// indicating the index of the bucket to which the measurement applies.
+func recordWithBucketIndex(ctx context.Context, bucketIndex int, ms ...stats.Measurement) {
+	_ = stats.RecordWithTags(ctx,
+		[]tag.Mutator{tag.Upsert(keyBucketIndex, string(bucketIndex))},
```

Comment:
`tag.Upsert` is a performance killer, unfortunately, so the way that this helper operates results in a lot of extra allocations. I suggest creating a helper that does the same as `LocalContext` but for bucket index and gets called after the `bucketID` is defined, just once.